### DISCUSSION TITLE
Unrelated to the course -- txpool with Foundry

### DISCUSSION THREAD

**Discussion by RichardPalladino** [Cyfrin/foundry-full-course-cu#579] [Q&A]
Hey all.  If you're like me, you love the speed and elegance of Foundry.  Unfortunately, I haven't figured out how to do everything I want with it that I can do with Hardhat.

Most especially, I haven't figure out how to read and use the txpool data with Forge scripts.  I've read through the entire sourcecode of the forge-std library and didn't find anything within that code.  I was kind of expecting to find something in Vm.sol.

Does anyone know whether you can read current txpool data using a Forge script?

**Comment by theirrationalone** [comment]
Here again in Foundry, We use **Alchemy** to monitor transaction pool data.

But Foundry has **Broadcast**, Which provides all transaction information-**(data)** but only two states **FAILED** & **COMPLETED**.

Here we can't interact with pending transactions, as The Helper we use **ALCHEMY** again. But You can see how a transaction went through on a real/forked testnet using **Broadcast**  with its data.

**Reply by theirrationalone** [reply]
And Foundry is completely, as we can say a security analysis, security audit, Protocol & its security testing, tool. So as in context a transaction has only one source **FOUNDRY**, So why would we need a transaction's pool(txpool) data for testing purpose only. We're not creating a **DAPP** here. We are using **FOUNDRY** For Testing & ensuring Security of a Hybrid Smart Contract.

Please answer, Am i wrong here????

**Reply by RichardPalladino** [reply]
I understand that Foundry is excellent for security and testing.  Frankly, execution of some attacks may include front running.  There are white-hats who are known to front-run malicious bots.  Foundry's speed would give them an advantage.  

Sticking to one piece of security - i.e., the dev/test stage - and precluding any operational use-cases is rather limiting.  Alchemy is fine if you don't care about latency.  I roll my own nodes because I require the low latency that IPC affords.  Front running a malicious attack demands such speed.

I very much appreciate what Foundry has done and how pleasurable of a tool it is.  I just wish it could cover all my use-cases in defi and smart contract security.

**Reply by theirrationalone** [reply]
If you're the block producer then only need to structure the pool therefore you'll already have all data that everyone can verify once your node mined(if you're the last node)/validated all transactions inside your node and after get attached to blockchain.

and **BEST OF LUCK** 🍀 For Your **DEFI** Testing with **Foundry**, Such a Great Tool.

**Reply by RichardPalladino** [reply]
It is a great smart contract security tool.  Good for most testing purposes too.  Not good for operational security.  Not good for many defi use-cases (and therefore not good for all defi testing).
